site/app/models/gradeable/Gradeable.php

```diff
@@ -0,0 +1,446 @@
+ * @method void setSyllabusBucket($bucket);
+ * @method Component[] getComponents();
+ * @method string getTaInstructions();
+ * @method void setTaInstructions($instructions);
+ * @method string getAutogradingConfigPath();
+ * @method bool isVcs();
+ * @method void setVcs($use_vcs);
+ * @method string getVcsSubdirectory();
+ * @method void setVcsSubdirectory($subdirectory);
+ * @method bool isTeamAssignment();
+ * @method int getTeamSizeMax();
+ * @method void setTeamSizeMax($max_team_size);
+ * @method \DateTime getTeamLockDate();
+ * @method bool isTaGrading();
+ * @method void setTaGrading($use_ta_grading);
+ * @method bool isStudentView();
+ * @method void setStudentView($can_student_view);
+ * @method bool isStudentSubmit();
+ * @method void setStudentSubmit($can_student_submit);
+ * @method bool isStudentDownload();
+ * @method void setStudentDownload($can_student_download);
+ * @method bool isStudentAnyVersion();
+ * @method void setStudentAnyVersion($student_any_version);
+ * @method bool isPeerGrading();
+ * @method void setPeerGrading($use_peer_grading);
+ * @method int getPeerGradeSet();
+ * @method void setPeerGradeSet($grade_set)
+ * @method \DateTime getSubmissionOpenDate();
+ * @method \DateTime getSubmissionDueDate();
+ * @method int getLateDays();
+ * @method bool isLateSubmissionAllowed();
+ * @method void setLateSubmissionAllowed($allow_late_submission)
+ * @method float getPrecision();
+ * @method void setPrecision($grading_precision);
+ */
+class Gradeable extends AbstractModel
+{
+    /* Properties for all types of gradeables */
+
+    /** @property @var string The course-wide unique gradeable id */
+    protected $id = "";
+    /** @property @var string The gradeable's title */
+    protected $title = "";
+    /** @property @var string The instructions url to give to students */
+    protected $instructions_url = "";
+    /** @property @var int The type of gradeable */
+    protected $type = GradeableType::ELECTRONIC_FILE;
+    /** @property @var bool If the gradeable should be graded per registration section (true) or rotating sections(false) */
+    protected $grade_by_registration = true;
+    /** @property @var \DateTime The so-called 'TA Beta-Testing' date.  This is when the gradeable appears for TA's */
+    protected $ta_view_start_date = null;
+    /** @property @var \DateTime The date that graders may start grading */
+    protected $grade_start_date = null;
+    /** @property @var \DateTime The date that grades will be released to students */
+    protected $grade_released_date = null;
+    /** @property @var \DateTime The date after which only instructors may change grades (aka when grades are 'due') */
+    protected $grade_locked_date = null;
+    /** @property @var int The minimum user group that can grade this gradeable (1=instructor) */
+    protected $min_grading_group = 1;
+    /** @property @var string The syllabus classification of this gradeable */
+    protected $syllabus_bucket = "Homework";
```

Comment:
Enum for this?

Reply:
For now we're going to let it be any string, but in the future we may want to setup a `buckets` table in the database and validate that its one of those.